velox/row/UnsafeRowFast.cpp

```diff
@@ -27,12 +27,187 @@ int32_t alignBits(int32_t numBits) {
 int32_t alignBytes(int32_t numBytes) {
   return bits::roundUp(numBytes, 8);
 }
+
+// Origins from BigInteger#firstNonzeroIntNum.
+int32_t firstNonzeroIntNum(const std::vector<int32_t>& mag, int32_t length) {
+  int32_t i;
+  for (i = length - 1; i >= 0 && mag[i] == 0; --i) {
+    ;
+  }
+  return length - i - 1;
+}
+
+// Origins from BigInteger#getInt.
+int32_t getInt(
+    int32_t n,
+    int32_t sig,
+    const std::vector<int32_t>& mag,
+    int32_t length) {
+  if (n < 0) {
+    return 0;
+  }
+  if (n >= length) {
+    return sig < 0 ? -1 : 0;
+  }
+
+  int32_t magInt = mag[length - n - 1];
+  return (
+      sig >= 0 ? magInt
+               : (n <= firstNonzeroIntNum(mag, length) ? -magInt : ~magInt));
+}
+
+int32_t getBitCount(uint32_t i) {
+  static constexpr int kMaxBits = std::numeric_limits<uint64_t>::digits;
+  uint64_t num = static_cast<uint32_t>(i);
+  return bits::countBits(reinterpret_cast<uint64_t*>(&num), 0, kMaxBits);
+}
+
+// Origins from java side BigInteger#bitLength.
+int32_t
+getBitLength(int32_t sig, const std::vector<int32_t>& mag, int32_t len) {
+  int32_t n = -1;
+  if (len == 0) {
+    n = 0;
+  } else {
+    // Calculate the bit length of the magnitude.
+    int32_t magBitLength = ((len - 1) << 5) + 64 -
+        bits::countLeadingZeros(static_cast<uint64_t>(mag[0]));
+    if (sig < 0) {
+      // Check if magnitude is a power of two.
+      bool pow2 = (getBitCount((uint32_t)mag[0]) == 1);
+      for (int i = 1; i < len && pow2; ++i) {
+        pow2 = (mag[i] == 0);
+      }
+      n = (pow2 ? magBitLength - 1 : magBitLength);
+    } else {
+      n = magBitLength;
+    }
+  }
+  return n;
+}
+
+// Refer to BigInteger(byte[] val) to get the mag definition
+std::vector<uint32_t>
+convertMagArray(int64_t newHigh, uint64_t newLow, int32_t* size) {
+  std::vector<uint32_t> mag;
```

Comment:
It feels inefficient to create std::vector for each value. Is this necessary? It looks like the size of this vector is always 4, so it could be an std::array, no?

Have you had a chance to look at the performance of this implementation? What do you see in the profile? How much slower long-decimal serde is compared to short-decimal?

Reply:
I optimize this function, and the performance is following when VectorFuzzer::Options  opts.vectorSize = 1000000;
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/cjin/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/cjin/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{border:.5pt solid windowtext;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">



unit: s | short   decimal(18, 18) | long   decimal(38, 18)
-- | -- | --
serialize | 0.13 | 1.35
deserialize | 0.37 | 0.49



</body>

</html>

Reply:
@jinchengchenghh It looks like serialization of long decimals is 10x slower than short decimals. Is this expected? Where does the CPU time go?

Reply:
I find converting from mag to byte array has serious performance issue, the new commit fixes this issue, and the new serialize long decimal time is 0.146624 which is similar to short decimal.

Reply:
@jinchengchenghh This is a nice finding. Thank you for the investigation. Glad to hear you were able to optimize.

Would it make sense to add a benchmark to velox/row/benchmark/UnsafeRowSerializeBenchmark.cpp ?